[CSS] Page zoom affects all frames devicePixelRatio equally.

mdubet opened the pull request:
#### bdfffc7361a623dc84c8c872d96f57bcbd458481
<pre>
[CSS] Page zoom affects all frames devicePixelRatio equally.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301807">https://bugs.webkit.org/show_bug.cgi?id=301807</a>
<a href="https://rdar.apple.com/163857955">rdar://163857955</a>

Reviewed by Simon Fraser.

This patch aligns behavior between all frames by making page zoom also affect the main frame
(it was already observable for iframe through usedZoom)

Test: fast/css/devicepixelratio-sync-iframe-page-zoom.html
* LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom-expected.txt: Added.
* LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom.html: Added.

New test to verify that main frame and iframe has the same DPR with page zoom.

* LayoutTests/fast/transforms/bounding-rect-zoom.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/zoom-iframe-dynamic.html:
* LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-2-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-huge-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-override-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-percentage-size-expected.txt:
* LayoutTests/svg/zoom/page/zoom-svg-as-object-expected.txt:

Rebaseline tests (they are correct)

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::frameScaleFactor const):

Canonical link: <a href="https://commits.webkit.org/303397@main">https://commits.webkit.org/303397@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e94b1c86046c70fcaa1c3ff5d9c2d5ec817df5bf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132295 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4790 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43331 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139810 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/351d6452-b2dd-4279-b4cd-0ceaf87e2f9a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134165 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4773 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4549 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101135 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b410ae78-52b0-4306-a5c3-6483106e7249) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135241 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118508 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81925 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3254 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83033 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142460 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4460 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37211 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109514 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4542 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109692 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27783 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3380 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114780 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57733 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4514 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4346 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67964 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4605 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4471 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom.html:
```diff
@@ -0,0 +1,41 @@
+        const mainDpiAfterZoom = window.devicePixelRatio;
+        const iframeDpiAfterZoom = iframe.contentWindow.devicePixelRatio;
+
+        assert_equals(
```
This tests whether they are the same, but not whether they are affected by zoom. We need that test too,

mdubet replied:
Yes indeed. The goal of this PR is *only* to verify that they are affected similarly by page zoom (whatever the effect).

mdubet replied:
I've added 2 assert to verify that they are affected by page zoom.

nt1m commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1132,7 +1132,7 @@ float LocalFrame::frameScaleFactor() const
         if (RefPtr ownerElement = this->ownerElement()) {
             if (auto* ownerRenderer = ownerElement->renderer())
-                return ownerRenderer->style().usedZoom();
+                return ownerRenderer->style().zoom();
```
Is this correct in the following case? In this case, iframe contents should be zoomed by 2. RenderStyle::usedZoom takes ancestor zoom in account, but is it the case for RenderStyle::zoom?

```
<div style="zoom: 2">
    <iframe src="https://ntim.me"></iframe>
</div>
```

data:text/html,<div%20style="zoom:%202"><iframe%20src="https://ntim.me">

mdubet replied:
You are right, we need to use the `usedZoom()`

smfr commented on LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom.html:
```diff
@@ -0,0 +1,50 @@
+        if (!window.internals) return;
+
+        // Initial state - should match
+        const initialMainDpi = window.devicePixelRatio;
```
DPR, not DPI

smfr commented on LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom.html:
```diff
@@ -0,0 +1,50 @@
+        assert_equals(
+            initialMainDpi,
+            initialIframeDpi,
+            "Initial DPI should match between main frame and iframe",
```
DPR, or spell it out.

smfr commented on LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom.html:
```diff
@@ -0,0 +1,50 @@
+        assert_equals(
+            mainDpiAfterZoom,
+            initialMainDpi *2,
+            "Main frame is zoomed",
```
The text should be `devicePixelRatio changes with zoom` or some such.

smfr commented on LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt:
```diff
@@ -31,8 +31,8 @@ layer at (0,0) size 800x355
                 layer at (0,0) size 133x29
-                  RenderSVGRoot {svg} at (0,0) size 93x21
-                    RenderSVGEllipse {circle} at (0,0) size 93x21 [stroke={[type=SOLID] [color=#000000]}] [fill={[type=SOLID] [color=#D9BB7A] [fill rule=EVEN-ODD]}] [cx=110.00] [cy=110.00] [r=110.00]
+                  RenderSVGRoot {svg} at (0,0) size 133x29
```
Do these tests look correct with the PR after zooming? Is it OK that all the SVG geometry is changing?

mdubet committed:
[CSS] Page zoom affects all frames devicePixelRatio equally.
https://bugs.webkit.org/show_bug.cgi?id=301807
rdar://163857955

Reviewed by Simon Fraser.

This patch aligns behavior between all frames by making page zoom also affect the main frame
(it was already observable for iframe through usedZoom)

Test: fast/css/devicepixelratio-sync-iframe-page-zoom.html
* LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom-expected.txt: Added.
* LayoutTests/fast/css/devicepixelratio-sync-iframe-page-zoom.html: Added.

New test to verify that main frame and iframe has the same DPR with page zoom.

* LayoutTests/fast/transforms/bounding-rect-zoom.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/zoom-iframe-dynamic.html:
* LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-2-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-absolute-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-huge-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-override-size-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-percentage-size-expected.txt:
* LayoutTests/svg/zoom/page/zoom-svg-as-object-expected.txt:

Rebaseline tests (they are correct)

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::frameScaleFactor const):

Canonical link: https://commits.webkit.org/303397@main
